projects/drivers/xilinx_platform/irq.c

```diff
@@ -0,0 +1,219 @@
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdlib.h>
+#include <xscugic.h>
+#include "error.h"
+#include "irq.h"
+#include "xilinx_platform_drivers.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Initialize the IRQ interrupts.
+ * @param desc - The IRQ descriptor.
+ * @param init_param - The structure that contains the IRQ parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t irq_ctrl_init(struct irq_desc **desc,
+		      const struct irq_init_param *param)
+{
+	int32_t status;
+	XScuGic_Config *IntcConfig;
+	XScuGic *gic;
+	struct xil_irq_desc *xil_dev;
+
+	struct irq_desc *descriptor = calloc(1, sizeof(struct irq_desc));
+	if (!descriptor)
+		return FAILURE;
+
+	descriptor->extra = calloc(1, sizeof(struct xil_irq_desc));
+	if(!(descriptor->extra)) {
+		free(descriptor);
+		return FAILURE;
+	}
+
+	gic = calloc(1, sizeof(XScuGic));
+	if (!gic) {
+		free(descriptor->extra);
+		free(descriptor);
+		return FAILURE;
+	}
+
+	xil_dev = descriptor->extra;
+	xil_dev->gic = gic;
+	/* Initialize the interrupt controller driver */
+	IntcConfig = XScuGic_LookupConfig(param->irq_id);
+	if (NULL == IntcConfig) {
+		free(gic);
```

Comment:
these appear as duplicates in the code. It could be done with goto error_free_gic. Something like this:

`error_free_gic:
          free(gic);
error_free_extra:
          free(descriptor->extra);
error_free_descriptor:
         free(descriptor); `